File: src/jp/ojt/sst/file/StackTraceFile.java
package jp.ojt.sst.file;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import jp.ojt.sst.exception.ASTException;
import jp.ojt.sst.model.StackTraceData;

import java.io.BufferedReader;
import java.io.IOException;

/**
 * StackTraceFile
 *
 */
public class StackTraceFile {

	/** Regex of date(log4j %d) */
	private static final String LEGEX_DATE = "(\\d{4}-\\d{2}-\\d{2})";
	
	/** Regex of Exception(contain package) */
	private static final String LEGEX_EXCEPTION="([\\w\\.]*Exception)";
	
	/** Target StackTraceLog absolute path */
	private String filePath;
	
	/** Search word into stackTraceLog */
	private String searchWord;
	
	/** Store StackTraceData.(date, Exception, detail, searchWord line) */
	private HashMap<String, StackTraceData> map = new HashMap<>();
	
	/**
	 * Constructor
	 * 
	 * @param path stackTraceLogFile absolute path
	 * @param word search word
	 */
	public StackTraceFile(String path, String word) {
		filePath = path;
		searchWord = word;
	}
	
	/**
	 * Read from Stack Trace File
	 */
	public void read() {
		
		Pattern datePtn = Pattern.compile(LEGEX_DATE);
		Pattern exceptionPtn = Pattern.compile(LEGEX_EXCEPTION);
		
		String matchDateStr = "";
		String exceptionStr = "";
		String messages = "";
		
		boolean foundDate = false;
		boolean foundException = false;
		boolean foundWord = false;
		
		try(BufferedReader br = Files.newBufferedReader(Paths.get(filePath))) {
			for(String line; (line = br.readLine()) != null; ) {
				
				Matcher dateMacher = datePtn.matcher(line);
				if(dateMacher.find()) {
					matchDateStr = dateMacher.group();
					foundDate = true;
					foundException = false;
					foundWord = false;
				}
				
				if(foundDate && !foundException) {
					Matcher exceptionMacher = exceptionPtn.matcher(line);
					if(exceptionMacher.find()) {
						exceptionStr = exceptionMacher.group();
						int idx = line.indexOf(exceptionStr) + exceptionStr.length();
						if (line.length() > idx + 1) {
							messages = line.substring(idx + 1);
						}
						foundException = true;
					}
				}
				
				if(foundException && !foundWord) {
					if(line.contains(searchWord)) {
						String key = matchDateStr + exceptionStr + messages + line;
						if(map.containsKey(key)) {
							StackTraceData stData = map.get(key);
							stData.addCount();
							map.replace(key, stData);
						} else {
							//TODO
							//StackTraceData stData = new StackTraceData(matchDateStr, exceptionStr, messages, line);
							//map.put(key, stData);
						}
						foundWord = true;
					}
				}
			}
		} catch (IOException ioe) {
			throw new ASTException(ioe);
		}
	}
	
	/**
	 * 
	 * @return summary stackTraceLogFile Map
	 */
	public HashMap<String, StackTraceData> getResultMap() {
		return map;
	}
}

File: src/jp/ojt/sst/file/StackTraceFile.kt
package jp.ojt.sst.file

import java.io.File;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.io.BufferedReader;
import java.io.IOException;

import jp.ojt.sst.model.StackTraceData;
import kotlin.io.*;
import kotlin.text.*;

class StackTraceFile(path: String, word: String) {
	
	val filePath = path
	val searchWord = word
	
	val LEGEX_DATE: String = "(\\d{4}-\\d{2}-\\d{2})"
	val LEGEX_EXCEPTION: String = "([\\w\\.]*Exception)"
	var map = HashMap<String, StackTraceData>()
	
	fun read() {
		val dateRegex = Regex(LEGEX_DATE)
		val exceptionRegex = Regex(LEGEX_EXCEPTION);
		
		var matchDateStr = ""
		var exceptionStr = ""
		var messages = ""
		
		var foundDate = false
		var foundException = false
		var foundWord = false
		
		File(filePath).absoluteFile.forEachLine { line ->
				
			if(dateRegex.containsMatchIn(line)) {
				matchDateStr = dateRegex.find(line)!!.value
				foundDate = true;
				foundException = false;
				foundWord = false;
			}
			
			if(foundDate && !foundException) {
				if(exceptionRegex.containsMatchIn(line)) {
					exceptionStr = exceptionRegex.find(line)!!.value
					val idx = line.indexOf(exceptionStr) + exceptionStr.length
					if (line.length > idx + 1) {
						messages = line.substring(idx + 1);
					}
					foundException = true;
				}
			}
				
			if(foundException && !foundWord) {
				if(line.contains(searchWord)) {
					val key = matchDateStr + exceptionStr + messages + line;
					if(map.containsKey(key)) {
						var stData: StackTraceData = map.get(key)!!
						stData.addCount();
						map.replace(key, stData);
					} else {
						var stData = StackTraceData(matchDateStr, exceptionStr, messages, line);
						map.put(key, stData);
					}
					foundWord = true;
				}
			}
		}
	}
}